vsphere/resource_vsphere_virtual_machine_test.go

```diff
@@ -1,1726 +1,1548 @@
+	// speeds up tests greatly, but it's not a necessary variable, so we don't
+	// enforce it here.
+	if os.Getenv("VSPHERE_DATACENTER") == "" {
+		t.Skip("set VSPHERE_DATACENTER to run vsphere_virtual_machine acceptance tests")
+	}
+	if os.Getenv("VSPHERE_CLUSTER") == "" {
+		t.Skip("set VSPHERE_CLUSTER to run vsphere_virtual_machine acceptance tests")
+	}
+	if os.Getenv("VSPHERE_RESOURCE_POOL") == "" {
+		t.Skip("set VSPHERE_RESOURCE_POOL to run vsphere_virtual_machine acceptance tests")
+	}
+	if os.Getenv("VSPHERE_NETWORK_LABEL") == "" {
+		t.Skip("set VSPHERE_NETWORK_LABEL to run vsphere_virtual_machine acceptance tests")
+	}
+	if os.Getenv("VSPHERE_IPV4_ADDRESS") == "" {
+		t.Skip("set VSPHERE_IPV4_ADDRESS to run vsphere_virtual_machine acceptance tests")
+	}
+	if os.Getenv("VSPHERE_IPV4_PREFIX") == "" {
+		t.Skip("set VSPHERE_IPV4_PREFIX to run vsphere_virtual_machine acceptance tests")
+	}
+	if os.Getenv("VSPHERE_IPV4_GATEWAY") == "" {
+		t.Skip("set VSPHERE_IPV4_GATEWAY to run vsphere_virtual_machine acceptance tests")
+	}
+	if os.Getenv("VSPHERE_DATASTORE") == "" {
+		t.Skip("set VSPHERE_DATASTORE to run vsphere_virtual_machine acceptance tests")
+	}
+	if os.Getenv("VSPHERE_TEMPLATE") == "" {
+		t.Skip("set VSPHERE_TEMPLATE to run vsphere_virtual_machine acceptance tests")
+	}
+}
 
-	if v := os.Getenv("VSPHERE_TEMPLATE"); v == "" {
-		t.Fatal("env variable VSPHERE_TEMPLATE must be set for acceptance tests")
+func testAccResourceVSphereVirtualMachineCheckExists(expected bool) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		_, err := testGetVirtualMachine(s, "vm")
+		if err != nil {
+			if ok, _ := regexp.MatchString("virtual machine with UUID \"[-a-f0-9]+\" not found", err.Error()); ok && !expected {
+				// Expected missing
+				return nil
+			}
+			return err
+		}
+		if !expected {
+			return errors.New("expected VM to be missing")
+		}
+		return nil
 	}
+}
 
-	if v := os.Getenv("VSPHERE_IPV4_GATEWAY"); v == "" {
-		t.Fatal("env variable VSPHERE_IPV4_GATEWAY must be set for acceptance tests")
+// testAccResourceVSphereVirtualMachineCheckPowerState is a check to check for
+// a VirtualMachine's power state.
+func testAccResourceVSphereVirtualMachineCheckPowerState(expected types.VirtualMachinePowerState) resource.TestCheckFunc {
+	return func(s *terraform.State) error {
+		props, err := testGetVirtualMachineProperties(s, "vm")
+		if err != nil {
+			return err
+		}
+		actual := props.Runtime.PowerState
+		if expected != actual {
+			return fmt.Errorf("expected power state to be %s, got %s", expected, actual)
+		}
+		return nil
 	}
+}
 
-	if v := os.Getenv("VSPHERE_IPV4_ADDRESS"); v == "" {
-		t.Fatal("env variable VSPHERE_IPV4_ADDRESS must be set for acceptance tests")
+// testAccResourceVSphereVirtualMachineCheckHostname is a check to check for a
+// VirtualMachine's hostname. The check uses guest info, so VMware tools needs
+// to be installed.
```

Comment:
What would happen if VMware tools weren't installed? Is there a way to check if the tools are installed before running tests that call this?

Reply:
If VMware tools is not installed, I think this test will either fail due to no data being present or panic, depending on how well populated `Guest` is populated in this scenario. In any case, it will fail, which is what we want it to do.

A good chunk of the resource right now really depends on VMware tools being installed on the template that you are basing it off of, and all of our tests (and most of the old tests as well) depend on a template. There might be a scenario in the future where we have a good use case where we need to deploy something with Terraform without VMware tools, like a virtual appliance - when we have that, we will probably need to create an environment variable that can be used to skip tests like these... but for now we don't have a test that would consume it, so it's not really necessary.